### PR TITLE
Adds apoc.metrics.storage(dirName) to allow inspection of free storage

### PR DESCRIPTION
**Pull request by moxious**
Adds apoc.metrics.storage(dirName) to allow inspection of free storage

## Proposed Changes (Mandatory)

  - A common operational problem with Neo4j is the disk filling up. When this happens, the database can't commit transactions and make them durable, which results in run-time "freak outs".  :)
  - Existing JMX metrics do not provide a free disk space metric to tell if this is going to happen.
  - Halin would like to be able to know if this is going to happen to warn the user as part of a diagnostic.

I've opted to implement this purely on a setting basis.  Users will have no external knowledge of where the mount points on disk are.  They will however be able to tell that the spot where they've put their TX logs is 99% full, and that would be a potential problem!

Reference: https://github.com/moxious/halin/issues/30

### CODE REVIEW

**Comment by moxious** [comment]
As with my last PR, unfortunately I can't provide tests for this because the ephemeral in memory graph database used for unit testing has no valid directory configurations or filesystem presence (I tried).  As part of containerized testing, I can provide later tests.

File: src/main/java/apoc/metrics/Metrics.java [hunk 1]
```diff
@@ -92,6 +95,57 @@ public Neo4jMetric(String name, long lastUpdated) {
...
+     * Note that we divulge the storage directory setting, not the internal path.
+     * It isn't a secret that neo4j has a dbms.directories.logs, but user doesn't need
+     * to know where that is on a physical hard disk.
+     */
+    public static class StorageMetric {
+        public final String setting;
+        public final long freeSpaceBytes;
+        public final long totalSpaceBytes;
+        public final long usableSpaceBytes;
+        public final double pctFree;
```
Comment (jexp) [inline src/main/java/apoc/metrics/Metrics.java]: perhaps spell it out?

File: src/main/java/apoc/metrics/Metrics.java [hunk 2]
```diff
@@ -92,6 +95,57 @@ public Neo4jMetric(String name, long lastUpdated) {
...
+
+        public StorageMetric(String setting, long freeSpaceBytes, long totalSpaceBytes, long usableSpaceBytes) {
+            this.setting = setting;
+            this.freeSpaceBytes = freeSpaceBytes;
+            this.totalSpaceBytes = totalSpaceBytes;
+            this.usableSpaceBytes = usableSpaceBytes;
+            this.pctFree = (double)freeSpaceBytes / (double)totalSpaceBytes;
```
Comment (jexp) [inline src/main/java/apoc/metrics/Metrics.java]: avoid division by zero?

File: src/main/java/apoc/metrics/Metrics.java [hunk 3]
```diff
@@ -92,6 +95,57 @@ public Neo4jMetric(String name, long lastUpdated) {
...
+        }
+    }
+
+    /**
+     * DAO that gets streamed to the user with apoc.metric.storage()
+     * Note that we divulge the storage directory setting, not the internal path.
+     * It isn't a secret that neo4j has a dbms.directories.logs, but user doesn't need
+     * to know where that is on a physical hard disk.
+     */
+    public static class StorageMetric {
```
Comment (jexp) [inline src/main/java/apoc/metrics/Metrics.java]: Cool that we don't expose paths.

File: src/main/java/apoc/metrics/Metrics.java [hunk 4]
```diff
@@ -125,6 +179,50 @@ public Neo4jMetric(String name, long lastUpdated) {
...
+            "as a directorySetting, you will get back all available directory settings.  For a list of available directory settings, " +
+            "see the Neo4j operations manual reference on configuration settings.   Directory settings are **not** paths, they are " +
+            "a neo4j.conf setting key name")
+    public Stream<StorageMetric> storage(@Name("directorySetting") String directorySetting) {
+        // Permit case-insensitive checks.
+        String input = directorySetting == null ? null : directorySetting.toLowerCase();
+
+        boolean validSetting = (
+            input == null ||
+            Arrays.stream(neo4jDirectoryConfigurationSettingNames).anyMatch(input::equals)
```
Comment (jexp) [inline src/main/java/apoc/metrics/Metrics.java]: contains?

File: src/main/java/apoc/metrics/Metrics.java [hunk 5]
```diff
@@ -125,6 +179,50 @@ public Neo4jMetric(String name, long lastUpdated) {
...
+                        return directorySetting.equals(dirSetting);
+                    }
+
+                    return true;
+                })
+                .map(StoragePair::fromDirectorySetting)
+                .filter(sp -> {
+                    if (sp == null) { return false; }
+
+                    if (!sp.dir.exists() || !sp.dir.isDirectory() || !sp.dir.canRead()) {
```
Comment (jexp) [inline src/main/java/apoc/metrics/Metrics.java]: can we invert the condition so it's more readable?

File: src/main/java/apoc/metrics/Metrics.java [hunk 6]
```diff
@@ -125,6 +179,50 @@ public Neo4jMetric(String name, long lastUpdated) {
...
+        if (!validSetting) {
+            String validOptions = String.join(", ", neo4jDirectoryConfigurationSettingNames);
+            throw new RuntimeException("Invalid directory setting specified.  Valid options are one of: " +
+                    validOptions);
+        }
+
+        return Stream.of(neo4jDirectoryConfigurationSettingNames)
+                .filter(dirSetting -> {
+                    // If user specified a particular one, immediately cut list to just that one.
+                    if (directorySetting != null) {
```
Comment (jexp) [inline src/main/java/apoc/metrics/Metrics.java]: `directorySetting == null || directorySetting.equals(dirSetting)`

File: src/main/java/apoc/metrics/Metrics.java [hunk 7]
```diff
@@ -204,4 +302,24 @@ private final double doubleFrom(LoadCsv.CSVResult r, String column) {
...
     }
+
+
+    // This is the list of dbms.directories.* valid configuration items for neo4j.
+    // https://neo4j.com/docs/operations-manual/current/reference/configuration-settings/
+    // Usually these reside under the same root but because they're separately configurable, in the worst case
+    // every one is on a different device.
+    //
+    // More likely, they'll be largely similar metrics.
+    public static final String [] neo4jDirectoryConfigurationSettingNames = new String [] {
```
Comment (jexp) [inline src/main/java/apoc/metrics/Metrics.java]: constants should be NEO4J_DIRECTORY_CONFIGURATION_SETTING_NAMES
use Arrays.asList(...) then you can use contains / stream directly on it.

**Comment by jexp** [comment]
Testing would be interesting with testcontainers :)
And docs entry.

File: src/main/java/apoc/metrics/Metrics.java (hunk 4 above)
Comment (moxious) [reply src/main/java/apoc/metrics/Metrics.java]: Don't think so...no.   If the input contains the valid setting, this means you could ask for storage metrics for "dbms.directories.logs@)#&*)&!)#$&!)#&$!#" --- not sure that makes sense, isn't it better to give an error in this case.

Comment (jexp) [reply src/main/java/apoc/metrics/Metrics.java]: contains as in list.contains(setting) not string contains :)

**Pull request merged**